### Backward Elimination

In [2]:
from warnings import filterwarnings
filterwarnings("ignore")

### Step 1 - Data Ingestion

In [3]:
import pandas as pd
df = pd.read_csv("training_set.csv", na_values=["","NA"], keep_default_na=False)
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

### Target Feature - SalePrice

### Step 2 - Perform basic data quality checks

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [5]:
m = df.isna().sum()
m

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [6]:
m[m>0]

LotFrontage      259
Alley           1369
MasVnrType         8
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

### Step 3 - Seperate X and Y(SalePrice)
Id is stastically insignificant

In [8]:
X = df.drop(columns=["Id","SalePrice"])
Y = df[["SalePrice"]]

In [9]:
X.head()

MSSubClass MSZoning  LotFrontage  ...  YrSold SaleType SaleCondition
0          60       RL         65.0  ...    2008       WD        Normal
1          20       RL         80.0  ...    2007       WD        Normal
2          60       RL         68.0  ...    2008       WD        Normal
3          70       RL         60.0  ...    2006       WD       Abnorml
4          60       RL         84.0  ...    2008       WD        Normal

[5 rows x 79 columns]

In [10]:
Y.head()

SalePrice
0     208500
1     181500
2     223500
3     140000
4     250000

### Step 4 - Create a preprocessing pipeline for feature selection
Categorical - Ordinal Encoding

In [11]:
cat = list(X.columns[X.dtypes == "object"])
con = list(X.columns[X.dtypes != "object"])

In [12]:
cat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [13]:
con

['MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold']

### Below pipeline is used for feature selection

In [14]:
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OrdinalEncoder
from sklearn.compose import ColumnTransformer

In [15]:
num_pipe1 = make_pipeline(
    SimpleImputer(strategy="mean"),
    SimpleImputer()
)

In [16]:
cat_pipe1 = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OrdinalEncoder()
)

In [17]:
pre1 = ColumnTransformer(
    [
        ("num", num_pipe1, con),
        ("cat", cat_pipe1, cat)
    ]
).set_output(transform="pandas")

In [18]:
X_pre = pre1.fit_transform(X)
X_pre.head()

num__MSSubClass  num__LotFrontage  ...  cat__SaleType  cat__SaleCondition
0             60.0              65.0  ...            8.0                 4.0
1             20.0              80.0  ...            8.0                 4.0
2             60.0              68.0  ...            8.0                 4.0
3             70.0              60.0  ...            8.0                 0.0
4             60.0              84.0  ...            8.0                 4.0

[5 rows x 79 columns]

### Apply Feature Selection

In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector

In [20]:
base_model = LinearRegression()

sel = SequentialFeatureSelector(base_model, direction="backward", n_features_to_select="auto")

sel.fit(X_pre, Y)

SequentialFeatureSelector(direction='backward', estimator=LinearRegression())

In [21]:
imp_cols = sel.get_feature_names_out()
imp_cols

array(['num__MSSubClass', 'num__LotArea', 'num__OverallQual',
       'num__OverallCond', 'num__YearBuilt', 'num__MasVnrArea',
       'num__BsmtFinSF1', 'num__BsmtFinSF2', 'num__BsmtUnfSF',
       'num__TotalBsmtSF', 'num__1stFlrSF', 'num__2ndFlrSF',
       'num__LowQualFinSF', 'num__GrLivArea', 'num__BsmtFullBath',
       'num__KitchenAbvGr', 'num__TotRmsAbvGrd', 'num__Fireplaces',
       'num__GarageCars', 'num__WoodDeckSF', 'num__OpenPorchSF',
       'num__ScreenPorch', 'num__PoolArea', 'num__YrSold',
       'cat__LandContour', 'cat__Utilities', 'cat__Neighborhood',
       'cat__BldgType', 'cat__HouseStyle', 'cat__RoofMatl',
       'cat__Exterior1st', 'cat__MasVnrType', 'cat__ExterQual',
       'cat__BsmtQual', 'cat__BsmtCond', 'cat__BsmtExposure',
       'cat__HeatingQC', 'cat__KitchenQual', 'cat__Functional',
       'cat__GarageCond'], dtype=object)

In [22]:
len(imp_cols)

40

In [23]:
sel_cols = [col.split("__")[1] for col in imp_cols]
sel_cols

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'OpenPorchSF',
 'ScreenPorch',
 'PoolArea',
 'YrSold',
 'LandContour',
 'Utilities',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'MasVnrType',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'GarageCond']

In [24]:
X_sel = X[sel_cols]
X_sel

MSSubClass  LotArea  OverallQual  ...  KitchenQual  Functional  GarageCond
0             60     8450            7  ...           Gd         Typ          TA
1             20     9600            6  ...           TA         Typ          TA
2             60    11250            7  ...           Gd         Typ          TA
3             70     9550            7  ...           Gd         Typ          TA
4             60    14260            8  ...           Gd         Typ          TA
...          ...      ...          ...  ...          ...         ...         ...
1455          60     7917            6  ...           TA         Typ          TA
1456          20    13175            6  ...           TA        Min1          TA
1457          70     9042            7  ...           Gd         Typ          TA
1458          20     9717            5  ...           Gd         Typ          TA
1459          20     9937            5  ...           TA         Typ          TA

[1460 rows x 40 columns]

### Feature Selection is done

### Create a final preprocessing pipeline on X_sel
Categorical - OneHotEncoding

In [25]:
cat_sel = list(X_sel.columns[X_sel.dtypes == "object"])
con_sel = list(X_sel.columns[X_sel.dtypes != "object"])

In [26]:
cat_sel

['LandContour',
 'Utilities',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'MasVnrType',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'GarageCond']

In [27]:
con_sel

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'OpenPorchSF',
 'ScreenPorch',
 'PoolArea',
 'YrSold']

In [28]:
from sklearn.preprocessing import OneHotEncoder

In [29]:
num_pipe2 = make_pipeline(
    SimpleImputer(strategy="mean"),
    StandardScaler()
)

In [30]:
cat_pipe2 = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OneHotEncoder(handle_unknown="ignore", sparse_output=False)
)

In [31]:
pre2 = ColumnTransformer(
    [
        ("num", num_pipe2, con_sel),
        ("cat", cat_pipe2, cat_sel)
    ]
).set_output(transform="pandas")

In [32]:
X_sel_pre = pre2.fit_transform(X_sel)
X_sel_pre.head()

num__MSSubClass  num__LotArea  ...  cat__GarageCond_Po  cat__GarageCond_TA
0         0.073375     -0.207142  ...                 0.0                 1.0
1        -0.872563     -0.091886  ...                 0.0                 1.0
2         0.073375      0.073480  ...                 0.0                 1.0
3         0.309859     -0.096897  ...                 0.0                 1.0
4         0.073375      0.375148  ...                 0.0                 1.0

[5 rows x 132 columns]

### Step 5 -Apply train test split

In [33]:
from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest = train_test_split(X_sel_pre, Y, test_size=0.2, random_state = 21)


In [34]:
xtrain.head()

num__MSSubClass  num__LotArea  ...  cat__GarageCond_Po  cat__GarageCond_TA
710         -0.636078     -0.640101  ...                 0.0                 1.0
1098        -0.163109     -0.452686  ...                 0.0                 1.0
1286        -0.872563     -0.072844  ...                 0.0                 1.0
992          0.073375     -0.075851  ...                 0.0                 1.0
631          1.492282     -0.593999  ...                 0.0                 1.0

[5 rows x 132 columns]

In [35]:
ytrain.head()

SalePrice
710       52000
1098     128000
1286     143000
992      187000
631      209500

In [36]:
xtest.head()

num__MSSubClass  num__LotArea  ...  cat__GarageCond_Po  cat__GarageCond_TA
880         -0.872563     -0.350058  ...                 0.0                 1.0
605          0.073375      0.309002  ...                 0.0                 1.0
1166        -0.872563     -0.004192  ...                 0.0                 1.0
216         -0.872563     -0.207142  ...                 0.0                 1.0
970         -0.163109      0.028380  ...                 0.0                 1.0

[5 rows x 132 columns]

In [37]:
ytest.head()

SalePrice
880      157000
605      205000
1166     245350
216      210000
970      135000

In [38]:
xtrain.shape

(1168, 132)

In [39]:
xtest.shape

(292, 132)

### Step 6 - Model Building

### 1. Linear Regression

In [40]:
model1 = LinearRegression()

model1.fit(xtrain, ytrain)

LinearRegression()

In [41]:
model1.intercept_

array([91202.5274584])

In [42]:
model1.coef_

array([[-3.41401086e+03,  4.57059085e+03,  8.20035944e+03,
         6.40761760e+03,  1.34978291e+04,  3.79609377e+03,
         6.92218732e+03,  3.90991193e+02, -1.32321958e+03,
         6.00764039e+03,  1.00004324e+04,  1.60458133e+04,
         4.84809386e+02,  2.07316422e+04,  3.88677707e+02,
        -1.93333041e+03,  1.02388138e+02,  7.30021748e+02,
         3.81459214e+03,  1.10549167e+03,  1.48152896e+03,
         1.79198453e+03,  3.27572509e+03, -2.13838580e+02,
        -2.91444325e+03,  5.50469355e+03, -3.67741853e+03,
         1.08716824e+03,  1.41778932e+04, -1.41778932e+04,
         7.49680971e+03, -4.59545022e+03,  5.74771968e+03,
         1.82122089e+03, -9.36700447e+03, -3.50770261e+03,
         1.36723663e+04, -1.05110000e+04, -6.57247476e+03,
        -8.45319535e+03, -9.69680376e+03, -1.43050109e+04,
        -1.31492617e+04,  1.20132998e+04, -1.43086814e+04,
         3.15838545e+04,  2.03393890e+04, -7.86395746e+03,
        -1.04352794e+04, -8.11012132e+03, -3.37491947e+0

### 2. Ridge

In [43]:
from sklearn.linear_model import Ridge

model2 = Ridge(alpha=1)

model2.fit(xtrain, ytrain)

Ridge(alpha=1)

In [44]:
model2.score(xtrain, ytrain)

0.9111209029450332

### Hyperparameter tuning

In [48]:
import numpy as np

params = {
    "alpha" : np.arange(start=0.1, stop=100, step=0.1)
}

params

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [50]:
from sklearn.model_selection import GridSearchCV

base_ridge = Ridge()

gscv_ridge = GridSearchCV(base_ridge, param_grid=params, cv=5, scoring="r2", n_jobs=-1)

gscv_ridge.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(), n_jobs=-1,
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,...
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [51]:
gscv_ridge.best_params_

{'alpha': np.float64(11.9)}

In [52]:
gscv_ridge.best_score_

np.float64(0.8383308721282836)

In [54]:
best_ridge = gscv_ridge.best_estimator_

best_ridge

Ridge(alpha=np.float64(11.9))

In [55]:
best_ridge.score(xtrain, ytrain)

0.8888832460016751

In [56]:
best_ridge.score(xtest, ytest)

0.83505474043859

### 3. Lasso

In [57]:
from sklearn.linear_model import Lasso

model3 = Lasso(alpha=1)

model3.fit(xtrain, ytrain)

Lasso(alpha=1)

In [58]:
model3.score(xtrain, ytrain)

0.9241407529819137

In [59]:
model3.score(xtest, ytest)

0.8178490322343647

### Hyperparameter tuning

In [64]:
params2 = {
    "alpha" : np.arange(start=0.1, stop=100, step=0.1)
}

params2

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [65]:
base_lasso = Lasso()

gscv_lasso = GridSearchCV(base_lasso, param_grid=params2, cv = 5, scoring="r2", n_jobs=-1)

gscv_lasso.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(), n_jobs=-1,
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,...
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [66]:
gscv_lasso.best_params_

{'alpha': np.float64(84.9)}

In [67]:
gscv_lasso.best_score_

np.float64(0.8479993569248462)

In [68]:
best_lasso = gscv_lasso.best_estimator_

In [69]:
best_lasso

Lasso(alpha=np.float64(84.9))

In [70]:
best_lasso.score(xtrain, ytrain)

0.9177945248791999

In [71]:
best_lasso.score(xtest, ytest)

0.8241715407526697

### Step 7 - Model evaluation

In [74]:
from sklearn.metrics import(
    root_mean_squared_error,
    mean_absolute_error,
    mean_absolute_percentage_error,
    r2_score
)

In [75]:
def evaluate_model(model, x, y):
    ypred = model.predict(x)
    rmse = root_mean_squared_error(y, ypred)
    mae = mean_absolute_error(y, ypred)
    mape = mean_absolute_percentage_error(y, ypred)
    r2 = r2_score(y, ypred)
    print(f"RMSE: {rmse:.2f}")
    print(f"MAE: {mae:.2f}")
    print(f"MAPE: {mape:.2%}")
    print(f"R2: {r2:.2%}")

In [82]:
print("Linear Regression Train Results")
evaluate_model(model1, xtrain, ytrain)

Linear Regression Train Results
RMSE: 21666.90
MAE: 14733.43
MAPE: 8.84%
R2: 92.41%


In [76]:
print("Linear Regression Test Results")
evaluate_model(model1, xtest, ytest)

Linear Regression Test Results
RMSE: 35201.27
MAE: 17415.04
MAPE: 10.25%
R2: 81.72%


In [85]:
print("Ridge Train Results")
evaluate_model(model2, xtrain, ytrain)

Ridge Train Results
RMSE: 23453.49
MAE: 15316.19
MAPE: 9.07%
R2: 91.11%


In [84]:
print("Ridge Test Results")
evaluate_model(model2, xtest, ytest)

Ridge Test Results
RMSE: 33090.73
MAE: 17235.63
MAPE: 10.10%
R2: 83.84%


In [86]:
print("Lasso Train Results")
evaluate_model(model3, xtrain, ytrain)

Lasso Train Results
RMSE: 21667.65
MAE: 14742.61
MAPE: 8.84%
R2: 92.41%


In [81]:
print("Lasso Test Results")
evaluate_model(model3, xtest, ytest)

Lasso Test Results
RMSE: 35135.37
MAE: 17351.70
MAPE: 10.21%
R2: 81.78%


### From above results, for given dataset Backward Elimination of feature selection has given better results. And Ridge giving Highest r2 score of 83.84%

### Step 8 - Out of Sample Prediction

In [88]:
xnew = pd.read_csv("sample_set.csv", na_values=["","NA"], keep_default_na=False)
xnew

Id  MSSubClass MSZoning  ...  YrSold  SaleType SaleCondition
0     1461          20       RH  ...    2010        WD        Normal
1     1462          20       RL  ...    2010        WD        Normal
2     1463          60       RL  ...    2010        WD        Normal
3     1464          60       RL  ...    2010        WD        Normal
4     1465         120       RL  ...    2010        WD        Normal
...    ...         ...      ...  ...     ...       ...           ...
1454  2915         160       RM  ...    2006        WD        Normal
1455  2916         160       RM  ...    2006        WD       Abnorml
1456  2917          20       RL  ...    2006        WD       Abnorml
1457  2918          85       RL  ...    2006        WD        Normal
1458  2919          60       RL  ...    2006        WD        Normal

[1459 rows x 80 columns]

In [90]:
pre2

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['MSSubClass', 'LotArea', 'OverallQual',
                                  'OverallCond', 'YearBuilt', 'MasVnrArea',
                                  'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF',
                                  'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
                                  'LowQualFinSF', 'GrLivArea', 'BsmtFullBath',
                                  'KitchenAbvGr', 'TotRmsAbvGrd',...
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['LandContour', 'Utilities', 'Neighborhood',
                                  'BldgType', 'HouseStyle', 'RoofMatl',
                                  'Exterior1st', 'MasVnrType', 'ExterQual',
                                  'BsmtQual', 'BsmtCond', 'BsmtExposure',
                                  'HeatingQC', 'KitchenQual', 'Functional',
                                  'GarageCond'])])

In [91]:
xnew_pre = pre2.transform(xnew)
xnew_pre

num__MSSubClass  num__LotArea  ...  cat__GarageCond_Po  cat__GarageCond_TA
0           -0.872563      0.110763  ...                 0.0                 1.0
1           -0.872563      0.375850  ...                 0.0                 1.0
2            0.073375      0.332053  ...                 0.0                 1.0
3            0.073375     -0.054002  ...                 0.0                 1.0
4            1.492282     -0.552407  ...                 0.0                 1.0
...               ...           ...  ...                 ...                 ...
1454         2.438219     -0.859988  ...                 0.0                 1.0
1455         2.438219     -0.864197  ...                 0.0                 1.0
1456        -0.872563      0.950423  ...                 0.0                 1.0
1457         0.664586     -0.007600  ...                 0.0                 1.0
1458         0.073375     -0.089180  ...                 0.0                 1.0

[1459 rows x 132 columns]

In [92]:
preds = best_ridge.predict(xnew_pre)
preds

array([120085.30901649, 150283.39465832, 166011.46812282, ...,
       174985.73250423, 105481.59255947, 220839.46324431])

In [93]:
xnew['SalePrice'] = preds

In [94]:
xnew = xnew.round(2)
xnew

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition  SalePrice
0     1461          20       RH  ...        WD         Normal  120085.31
1     1462          20       RL  ...        WD         Normal  150283.39
2     1463          60       RL  ...        WD         Normal  166011.47
3     1464          60       RL  ...        WD         Normal  177817.02
4     1465         120       RL  ...        WD         Normal  199488.80
...    ...         ...      ...  ...       ...            ...        ...
1454  2915         160       RM  ...        WD         Normal   75963.10
1455  2916         160       RM  ...        WD        Abnorml   81459.23
1456  2917          20       RL  ...        WD        Abnorml  174985.73
1457  2918          85       RL  ...        WD         Normal  105481.59
1458  2919          60       RL  ...        WD         Normal  220839.46

[1459 rows x 81 columns]

In [95]:
Predictions = xnew[["Id","SalePrice"]]
Predictions

Id  SalePrice
0     1461  120085.31
1     1462  150283.39
2     1463  166011.47
3     1464  177817.02
4     1465  199488.80
...    ...        ...
1454  2915   75963.10
1455  2916   81459.23
1456  2917  174985.73
1457  2918  105481.59
1458  2919  220839.46

[1459 rows x 2 columns]

In [96]:
Predictions.to_csv("SalesPrice.csv",index=False)

In [97]:
pd.read_csv("SalesPrice.csv")

Id  SalePrice
0     1461  120085.31
1     1462  150283.39
2     1463  166011.47
3     1464  177817.02
4     1465  199488.80
...    ...        ...
1454  2915   75963.10
1455  2916   81459.23
1456  2917  174985.73
1457  2918  105481.59
1458  2919  220839.46

[1459 rows x 2 columns]